### PR TITLE
dns: parse additional RR types

### PR DESCRIPTION
**Pull request by satta**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: 
https://redmine.openinfosecfoundation.org/issues/3663

Describe changes:

The DNS parser code was extended to also parse the payloads for `NULL`, `NS` and `SRV` records. These could be useful to detect and document DNS tunneling activity, among other use cases.

New EVE-JSON examples:

```json
  "dns": {
    "version": 2,
    "type": "answer",
    "id": 1064,
    "flags": "8400",
    "qr": true,
    "aa": true,
    "rrname": "sbhi1c.pirate.sea",
    "rrtype": "NULL",
    "rcode": "NOERROR",
    "answers": [
      {
        "rrname": "sbhi1c.pirate.sea",
        "rrtype": "NULL",
        "ttl": 0,
        "rdata": "Base128"
      }
    ],
    "grouped": {
      "NULL": [
        "Base128"
      ]
    }
  }
```

```json
  "dns": {
    "version": 2,
    "type": "answer",
    "id": 6158,
    "flags": "81a0",
    "qr": true,
    "rd": true,
    "ra": true,
    "rrname": "isc.org",
    "rrtype": "NS",
    "rcode": "NOERROR",
    "answers": [
      {
        "rrname": "isc.org",
        "rrtype": "NS",
        "ttl": 7127,
        "rdata": "ns1.isc.org"
      },
      {
        "rrname": "isc.org",
        "rrtype": "NS",
        "ttl": 7127,
        "rdata": "ns3.isc.org"
      },
      {
        "rrname": "isc.org",
        "rrtype": "NS",
        "ttl": 7127,
        "rdata": "ns.isc.afilias-nst.info"
      },
      {
        "rrname": "isc.org",
        "rrtype": "NS",
        "ttl": 7127,
        "rdata": "ns2.isc.org"
      }
    ],
    "grouped": {
      "NS": [
        "ns1.isc.org",
        "ns3.isc.org",
        "ns.isc.afilias-nst.info",
        "ns2.isc.org"
      ]
    }
  }
```

```json
  "dns": {
    "version": 2,
    "type": "answer",
    "id": 60246,
    "flags": "8180",
    "qr": true,
    "rd": true,
    "ra": true,
    "rrname": "_sip._udp.sip.voice.google.com",
    "rrtype": "SRV",
    "rcode": "NOERROR",
    "answers": [
      {
        "rrname": "_sip._udp.sip.voice.google.com",
        "rrtype": "SRV",
        "ttl": 275,
        "srv": {
          "priority": 20,
          "weight": 1,
          "port": 5060,
          "name": "sip-anycast-2.voice.google.com",
          "raw": "\u0000\u0014\u0000\u0001\u0013\\xc4\rsip-anycast-2\u0005voice\u0006google\u0003com\u0000"
        }
      },
      {
        "rrname": "_sip._udp.sip.voice.google.com",
        "rrtype": "SRV",
        "ttl": 275,
        "srv": {
          "priority": 10,
          "weight": 1,
          "port": 5060,
          "name": "sip-anycast-1.voice.google.com",
          "raw": "\u0000\n\u0000\u0001\u0013\\xc4\rsip-anycast-1\u0005voice\u0006google\u0003com\u0000"
        }
      }
    ],
    "grouped": {
      "SRV": [
        {
          "priority": 20,
          "weight": 1,
          "port": 5060,
          "name": "sip-anycast-2.voice.google.com",
          "raw": "\u0000\u0014\u0000\u0001\u0013\\xc4\rsip-anycast-2\u0005voice\u0006google\u0003com\u0000"
        },
        {
          "priority": 10,
          "weight": 1,
          "port": 5060,
          "name": "sip-anycast-1.voice.google.com",
          "raw": "\u0000\n\u0000\u0001\u0013\\xc4\rsip-anycast-1\u0005voice\u0006google\u0003com\u0000"
        }
      ]
    }
  }
```

### CODE REVIEW

**Comment by satta** [comment]
No idea what's going on with that AppVeyor build.

**Comment by jasonish** [comment]
> No idea what's going on with that AppVeyor build.

Its not you, its them.

File: doc/userguide/output/eve/eve-json-format.rst [hunk 1]
```diff
@@ -422,6 +422,14 @@ More complex DNS record types may log additional fields for resource data:
...
   * "type": Fingerprint type (ex: 1 for SHA-1)
 
+* "srv": section containing fields for the SRV (location of services) record type
+
+  * "target": Domain name of the target host (ex: ``foo.bar.baz``)
+  * "priority": Target priority (ex: 20)
+  * "weight": Weight for target selection (ex: 1)
+  * "port": Port on this target host of this service (ex: 5060)
+  * "raw": Raw resource data bytes (e.g. for DNS tunneling detection)
+
```
Comment (jasonish) [inline doc/userguide/output/eve/eve-json-format.rst:432]: This doc change should be in its own commit please.

**Review by jasonish — CHANGES_REQUESTED** [review]
Also, did you see PRs #5446 and #5478 that are pending review?

File: rust/src/dns/lua.rs [hunk 2]
```diff
@@ -197,6 +199,11 @@ pub extern "C" fn rs_dns_lua_get_answer_table(clua: &mut CLuaState,
                     lua.pushstring(&String::from_utf8_lossy(&sshfp.fingerprint));
                     lua.settable(-3);
                 },
+                DNSRData::SRV(ref srv) => {
+                    lua.pushstring("addr");
+                    lua.pushstring(&String::from_utf8_lossy(&srv.target));
+                    lua.settable(-3);
+                },
```
Comment (jasonish) [inline rust/src/dns/lua.rs:206]: This should be squashed into the previous commit as the previous commit is unlikely to build without this.

**Comment by satta** [comment]
> Also, did you see PRs #5446 and #5478 that are pending review?

No, indeed I didn't. Hm. Should I remove the `NS` and `NULL` parts and just keep `SRV` in v2 to eliminate overlap between these PRs? I think that's the best idea... what do you think?

**Comment by jasonish** [comment]
> No, indeed I didn't. Hm. Should I remove the `NS` and `NULL` parts and just keep `SRV` in v2 to eliminate overlap between these PRs? I think that's the best idea... what do you think?

I've created PR https://github.com/OISF/suricata/pull/5528 that brings both of those into a single PR.  Can you rebase this work onto that? That really means just doing the SRV support.  Thanks!

**Comment by jasonish** [comment]
I should also add that the code looks fine, so more a commit fixup. Also, please provide your version of the null parser where you use rest(), I think thats the better function here.

File: doc/userguide/output/eve/eve-json-format.rst (hunk 1 above)
Comment (satta) [reply doc/userguide/output/eve/eve-json-format.rst:432]: Done.

File: rust/src/dns/lua.rs (hunk 2 above)
Comment (satta) [reply rust/src/dns/lua.rs:206]: Done.

**Comment by satta** [comment]
> I should also add that the code looks fine, so more a commit fixup. Also, please provide your version of the null parser where you use rest(), I think thats the better function here.

Done. See https://github.com/OISF/suricata/pull/5529
